services/src/main/java/org/keycloak/social/apple/AppleIdentityProvider.java

```diff
@@ -0,0 +1,277 @@
+    }
+
+    protected class Endpoint {
+        protected RealmModel realm;
+
+        protected AuthenticationCallback callback;
+
+        protected EventBuilder event;
+
+        @Context
+        protected KeycloakSession session;
+
+        @Context
+        protected ClientConnection clientConnection;
+
+        @Context
+        protected HttpHeaders headers;
+
+        public Endpoint(RealmModel realm, AuthenticationCallback callback, EventBuilder event) {
+            this.realm = realm;
+            this.callback = callback;
+            this.event = event;
+        }
+
+        @POST
+        public Response authResponse(@FormParam(AppleIdentityProvider.OAUTH2_PARAMETER_STATE) String state,
+                                     @FormParam(AppleIdentityProvider.OAUTH2_PARAMETER_CODE) String authorizationCode,
+                                     @FormParam(AppleIdentityProvider.OAUTH2_PARAMETER_USER) String user,
+                                     @FormParam(OAuth2Constants.ERROR) String error) {
+            if (error != null) {
+                logger.error(error + " for broker login " + getConfig().getProviderId());
+                if (error.equals(ACCESS_DENIED)) {
+                    return callback.cancelled(state);
+                } else if (error.equals(OAuthErrorException.LOGIN_REQUIRED) || error.equals(OAuthErrorException.INTERACTION_REQUIRED)) {
+                    return callback.error(state, error);
+                } else {
+                    return callback.error(state, Messages.IDENTITY_PROVIDER_UNEXPECTED_ERROR);
+                }
+            }
+
+            try {
+                if (authorizationCode != null) {
+
+                    String response = generateTokenRequest(authorizationCode).asString();
+
+                    BrokeredIdentityContext federatedIdentity = getFederatedIdentity(user, response);
+                    federatedIdentity.setIdpConfig(getConfig());
+                    federatedIdentity.setIdp(AppleIdentityProvider.this);
+                    federatedIdentity.setCode(state);
+                    return callback.authenticated(federatedIdentity);
+                }
+            } catch (WebApplicationException e) {
+                return e.getResponse();
+            } catch (Exception e) {
+                logger.error("Failed to make identity provider oauth callback", e);
+            }
+            event.event(EventType.LOGIN);
+            event.error(Errors.IDENTITY_PROVIDER_LOGIN_FAILURE);
+            return ErrorPage.error(session, null, Response.Status.BAD_GATEWAY, Messages.IDENTITY_PROVIDER_UNEXPECTED_ERROR);
+        }
+
+        public BrokeredIdentityContext getFederatedIdentity(String userData, String response) throws JsonProcessingException {
+            BrokeredIdentityContext user = AppleIdentityProvider.this.getFederatedIdentity(response);
+
+            if (userData != null) {
+                JsonNode profile = mapper.readTree(userData);
```

Comment:
Does that mean Apple returns user information through the frontend as a JSON string?

Reply:
yes the user information is returned as a JSON string by the Apple rest endpoint.